modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/npm/NPMRegistryResolutionStateDigestUtil.java

```diff
@@ -0,0 +1,204 @@
+ * share the same state, their digests are identical.
+ *
+ * @author Iván Zaera Avellón
+ * @review
+ */
+public class NPMRegistryResolutionStateDigestUtil {
+
+	public static String digest(NPMRegistry npmRegistry) {
+		MessageDigest messageDigest;
+
+		try {
+			messageDigest = MessageDigest.getInstance("SHA-1");
+		}
+		catch (NoSuchAlgorithmException noSuchAlgorithmException) {
+			throw new RuntimeException(noSuchAlgorithmException);
+		}
+
+		// Hash DynamicModules which don't honor immutability of packages
+		// with given (name, version) value
+
+		List<DynamicJSModule> dynamicJSModules = new ArrayList<>();
+
+		for (JSModule jsModule : npmRegistry.getResolvedJSModules()) {
+			if (jsModule instanceof DynamicJSModule) {
+				dynamicJSModules.add((DynamicJSModule)jsModule);
+			}
+		}
+
+		Collections.sort(
+			dynamicJSModules,
+			(dynamicJSModule1, dynamicJSModule2) -> {
+				String resolvedId = dynamicJSModule1.getResolvedId();
+
+				return resolvedId.compareTo(dynamicJSModule2.getResolvedId());
+			});
+
+		for (DynamicJSModule dynamicJSModule : dynamicJSModules) {
+			_update(messageDigest, dynamicJSModule);
+		}
+
+		// Now hash packages alone
+
+		List<JSPackage> jsPackages = new ArrayList<>(
+			npmRegistry.getResolvedJSPackages());
+
+		Collections.sort(
+			jsPackages,
+			(jsPackage1, jsPackage2) -> {
+				String resolvedId = jsPackage1.getResolvedId();
+
+				return resolvedId.compareTo(jsPackage2.getResolvedId());
+			});
+
+		for (JSPackage jsPackage : jsPackages) {
+			_update(messageDigest, jsPackage);
+		}
+
+		// Finally, hash the list of applied patches. This is necessary because
+		// Liferay Support's patches break the immutability convention of
+		// packages with given (name, version) value.
+
+		List<String> installedPatches = Arrays.asList(
+			PatcherUtil.getInstalledPatches());
+
+		Collections.sort(installedPatches);
+
+		for (String installedPatch : installedPatches) {
+			_update(messageDigest, installedPatch);
+		}
+
+		return StringUtil.bytesToHexString(messageDigest.digest());
+	}
+
+	private static void _update(
+		MessageDigest messageDigest, DynamicJSModule dynamicJSModule) {
+
+		_update(messageDigest, dynamicJSModule.getResolvedId());
+
+		List<String> dependencies = new ArrayList<>(
+			dynamicJSModule.getDependencies());
+
+		Collections.sort(dependencies);
+
+		for (String dependency : dependencies) {
+			_update(messageDigest, dependency);
+		}
+	}
+
+	private static void _update(
+		MessageDigest messageDigest, JSPackage jsPackage) {
+
+		// In theory name and version should be enough because npm packages are
+		// supposed to be immutable, but since it doesn't take too much longer,
+		// we add more fields to be safer.
+
+		_update(messageDigest, jsPackage.getMainModuleName());
+		_update(messageDigest, jsPackage.getName());
+		_update(messageDigest, jsPackage.getVersion());
+
+		List<JSPackageDependency> jsPackageDependencies = new ArrayList<>(
```

Comment:
I'm assuming we'll need this because we can't make sure package version will be increased every time a dependency changes. So as long as this takes short time I'm fine